extra/mariadb_upgrade_config_file.c

```diff
@@ -0,0 +1,1171 @@
+    const char *mariadbd_name= IF_WIN("mariadbd.exe", "mariadbd");
+
+    if (init_dynamic_string(&mariadbd_output, "", 32, 32))
+      goto err;
+    if (test_mariadbd(mariadbd_name))
+      goto err;
+    if ((error= run_mariadbd(mariadbd_name, &mariadbd_output, defaults_args, defaults_args_count)))
+    {
+      fputs("error: Failed to run mariadbd with the updated files, reverting\n", stderr);
+      if (error > 0)
+        fprintf(stderr, "mariadbd output:\n%s", mariadbd_output.str);
+    }
+    dynstr_free(&mariadbd_output);
+  }
+
+  error= finish_updated_files(&ctx, error == 0);
+  free_root(&alloc, MYF(0));
+  return error || ctx.failed;
+
+err:
+  finish_updated_files(&ctx, FALSE);
+  free_root(&alloc, MYF(0));
+  return 1;
+}
+
+
+static const char *config_file="my";			/* Default config file */
+
+enum upgrade_config_options
+{
+  OPT_UPDATE = 256,
+  OPT_BACKUP,
+  OPT_CURRENT_VERSION,
+  OPT_EDIT,
+  OPT_PRINT,
+};
+
+static struct my_option my_long_options[] =
+{
+  {"help", '?', "Display this help message and exit.",
+   0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0},
+  {"version", 'V', "Output version information and exit.",
+   0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0},
+  {"update", OPT_UPDATE, "Update the configuration files in place.",
+   0, 0, 0, GET_NO_ARG, NO_ARG, FALSE, 0, 0, 0, 0, 0},
+  {"backup", OPT_BACKUP,
+   "Backup the updated configuration files. The backup file names end in a "
+   "timestamp followed by .BAK",
+   0, 0, 0, GET_NO_ARG, NO_ARG, FALSE, 0, 0, 0, 0, 0},
+  {"current-version", OPT_CURRENT_VERSION,
+   "Section to use for invalid options. See --edit.",
+   &opt_current_version, &opt_current_version,
+   0, GET_STR, REQUIRED_ARG, 0, 0, 0, 0, 0, 0},
+  {"edit", OPT_EDIT,
+   "Select what to do with invalid options",
+   &opt_edit_mode, &opt_edit_mode, &edit_mode_typelib, GET_ENUM,
+   REQUIRED_ARG, EDIT_MODE_NONE, 0, 0, 0, 0, 0},
+  {"print", OPT_PRINT, "Print upgraded files to stdout.",
+   0, 0, 0, GET_NO_ARG, NO_ARG, FALSE, 0, 0, 0, 0, 0},
+  {0, 0, 0, 0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0}
+};
+
+static void cleanup_and_exit(int exit_code) __attribute__ ((noreturn));
+static void cleanup_and_exit(int exit_code)
+{
+  my_end(0);
+  exit(exit_code);
+}
+
+static void usage() __attribute__ ((noreturn));
+static void usage()
+{
+  print_version();
+  puts("This software comes with ABSOLUTELY NO WARRANTY. This is free software,\nand you are welcome to modify and redistribute it under the GPL license\n");
+  puts("Displays the unrecognized options present in configuration files, which is useful when upgrading MariaDB");
+  printf("Usage: %s [OPTIONS]\n", my_progname);
+  my_print_help(my_long_options);
+  my_print_default_files(config_file);
+  cleanup_and_exit(0);
+}
+
+
+static my_bool
+get_one_option(const struct my_option *opt __attribute__((unused)),
+	       const char *argument __attribute__((unused)),
+               const char *filename __attribute__((unused)))
```

Comment:
If this file had been C++ instead of C, you could simply write the following:
```c++
static my_bool get_one_option(const my_option*, const char*, const char*)
```